typhon/signal.py

```diff
@@ -56,29 +57,44 @@ def signal_widget(signal, read_only=False):
         # Register signal with plugin
         register_signal(signal)
         chan = channel_name(signal.name, protocol='sig')
-    # Check for enum_strs, if so create a QCombobox
-    if read_only:
-        logger.debug("Creating Label for %s", signal.name)
-        widget = TyphonLabel
-        name = signal.name + '_label'
-    else:
-        # Grab a description of the widget to see the correct widget type
-        try:
-            desc = signal.describe()[signal.name]
-        except Exception:
-            logger.error("Unable to connect to %r during widget creation",
-                         signal.name)
-            desc = {}
-        # Create a QCombobox if the widget has enum_strs
-        if 'enum_strs' in desc:
-            logger.debug("Creating Combobox for %s", signal.name)
-            widget = TyphonComboBox
-            name = signal.name + '_combo'
-        # Otherwise a LineEdit will suffice
+    # Grab a description of the widget to see the correct widget type
+    try:
+        desc = signal.describe()[signal.name]
+    except Exception:
+        logger.error("Unable to connect to %r during widget creation",
+                     signal.name)
+        desc = {}
+    # Unshaped data
+    if desc.get('shape', []) == []:
+        # Check for enum_strs, if so create a QCombobox
+        if read_only:
+            logger.debug("Creating Label for %s", signal.name)
+            widget = TyphonLabel
+            name = signal.name + '_label'
         else:
-            logger.debug("Creating LineEdit for %s", signal.name)
-            widget = TyphonLineEdit
-            name = signal.name + '_edit'
+            # Create a QCombobox if the widget has enum_strs
+            if 'enum_strs' in desc:
+                logger.debug("Creating Combobox for %s", signal.name)
+                widget = TyphonComboBox
+                name = signal.name + '_combo'
+            # Otherwise a LineEdit will suffice
+            else:
+                logger.debug("Creating LineEdit for %s", signal.name)
+                widget = TyphonLineEdit
+                name = signal.name + '_edit'
+    # Waveform
+    elif len(desc.get('shape')) == 1:
+        logger.debug("Creating WaveformDialogButton for %s", signal.name)
+        widget = WaveformDialogButton
+        name = signal.name + '_waveform_button'
```

Comment:
It's probably unnecessary to do this any more dynamically, but perhaps at least you could make a `{widget: suffix}` dictionary and create the signal name below this big if block? All these `debug` messages could also be wrapped into one message as `widget` and `name` give you the full string.

Reply:
I want to refactor this whole function. I think just a single row of the panel should be usable in designer. I also want to make sure that you can indicate via stylesheet if you want a specific type of widget there. I.e Show `low_limit_switch` as an indicator light.